Add MP3 decoder to fmt

Yanrishatum opened the pull request:
Introduces MP3 decoder support. Utilizes [lieft/minimp3](https://github.com/lieff/minimp3) library to decode mp3.
Exposes two functions:
- `mp3_open(bytes, size)` - allocates reader structure.
- `mp3_decode_frame(mp3, pos, output, size, offset)` - reads frame at `pos` and outputs to `output + offset`.

Notes/concerns/details:
* I use `Float32` output mode instead of `Short`.
* I don't like double-buffering instead of direct write to `output` on `mp3_decode_frame`.
* There is no checks if frame will fit into output and it may corrupt memory due to that.
* Probably worth adding ability to read multiple frames at once, since 44100 mp3 will read a bunch of frames at once to fill the audio buffer.

I'm not very confident with C code, and would appreciate comments on how to improve this implementation.
Related: Heaps MP3 support

ncannasse commented:
For the same reason as https://github.com/HeapsIO/heaps/pull/514, I would prefer this to be part of a standalone library.

ncannasse commented on libs/fmt/fmt.c:
```diff
@@ -418,6 +422,41 @@ DEFINE_PRIM(_I32, ogg_tell, _OGG);
+	ret = mp3dec_decode_frame(&o->dec, o->bytes + position, o->size - position, o->pcm, &o->info);
+	// TODO: Without memcpy?
+	// TODO: Safe write
+	memcpy( (void *)(output + offset), (void *)o->pcm, sizeof(o->pcm) );
```
I don't think sizeof(o->pcm) is the best here. You should only copy the amount of frame_bytes as specified in documentation. Please also handle errors (ret = 0) and return -1 without copying if the size to copy exceed `size`

Yanrishatum replied:
`dec.frame_bytes` refer to amount of bytes read from input buffer, e.g. assumed next frame offset. To copy only the amount of decoded samples (`pcm` uses 2-channel maximum size) it would need to calculate them trough `ret * o->info.channels * 4`.
As of errors:
`-1` when `position` is out of bounds of file bounds and when output buffer is too small.
`0` when no mp3 data was decoded.

Yanrishatum replied:
Then output size should be as big as pcm, because it's impossible to know the sample count in frame before decoding it (or doing extra work on finding and reading header and interpreting it to detect amount of samples)

ncannasse commented:
Looking at the minimal dependencies, I agree we can merge this without much concern, independently from the discussion in Heaps

ncannasse commented on libs/fmt/fmt.c:
```diff
@@ -418,6 +422,41 @@ DEFINE_PRIM(_I32, ogg_tell, _OGG);
+};
+
+HL_PRIM fmt_mp3 *HL_NAME(mp3_open)( char *bytes, int size ) {
+	fmt_mp3 *o = (fmt_mp3*)hl_gc_alloc_raw(sizeof(fmt_mp3));
```
You should not use allow_raw here, as this will make the GC scan the whole data for pointers, eventually creating false positive. What you want here is hl_gc_alloc_noptr, or alloc_finalizer if some memory needs to be freed when the decoder is no longer in-use.

Yanrishatum replied:
I'm not sure why, but if I use `alloc_noptr` or `alloc_finalizer` it throws `access_violation` when attempting to decode a frame.

ncannasse commented:
Both does not keep pointers so you have to keep a reference to bytes on
Haxe side so they don't get GCed.

Another option is to change the api so bytes gets passed for every decode
call.

Le sam. 5 janv. 2019 20:19, Pavel Alexandrov <notifications@github.com> a
écrit :

> *@Yanrishatum* commented on this pull request.
> ------------------------------
>
> In libs/fmt/fmt.c
> <https://github.com/HaxeFoundation/hashlink/pull/214#discussion_r245485954>
> :
>
> > @@ -418,6 +422,41 @@ DEFINE_PRIM(_I32, ogg_tell, _OGG);
>  DEFINE_PRIM(_BOOL, ogg_seek, _OGG _I32);
>  DEFINE_PRIM(_I32, ogg_read, _OGG _BYTES _I32 _I32);
>
> +/* ----------------------------------------------- SOUND : MP3 ------------------------------------------------ */
> +
> +typedef struct _fmt_mp3 fmt_mp3;
> +struct _fmt_mp3 {
> +	char *bytes;
> +	int size;
> +	mp3dec_t dec;
> +	mp3dec_frame_info_t info;
> +	mp3d_sample_t pcm[MINIMP3_MAX_SAMPLES_PER_FRAME];
> +};
> +
> +HL_PRIM fmt_mp3 *HL_NAME(mp3_open)( char *bytes, int size ) {
> +	fmt_mp3 *o = (fmt_mp3*)hl_gc_alloc_raw(sizeof(fmt_mp3));
>
> I'm not sure why, but if I use alloc_noptr or alloc_finalizer it throws
> access_violation when attempting to decode a frame.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/HaxeFoundation/hashlink/pull/214#discussion_r245485954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-bwMHoGMaln16pviodq_d9vwmg5SjNks5vAPrVgaJpZM4ZlkBp>
> .
>

Yanrishatum commented:
Changed API to require passing bytes each time (Which also allows for livestreaming of mp3 as it downloads from internet), added error handling, fixed gc_alloc and also added `mp3_frame_info` helper function.
